students/alex_skrn/lesson05/mailroom.py

```diff
@@ -0,0 +1,244 @@
+    print()
+    donors_L = list(donors.keys())
+    for name in donors_L[:-1]:
+        print(name, end=', ')
+    print(donors_L[-1])
+    print()
+
+
+def get_email(name, amount):
+    """Return a str containing a thank-you email."""
+    email_text = ("""\nDear {},\n
+                  \nI would like to thank you for your donation of ${}.\n
+                  \nWe appreciate your support.\n
+                  \nSincerely,\n
+                  \nThe Organization\n
+                  """)
+    return email_text.format(name, amount)
+
+
+def print_email(name, amount):
+    """Print a thank-you email on screen."""
+    print(get_email(name, amount))
+
+
+def create_report():
+    """Create a report."""
+    title_line_form = "{:<26}{:^3}{:>13}{:^3}{:>13}{:^3}{:>13}"
+    title_line_text = ('Donor Name', '|', 'Total Given', '|',
+                       'Num Gifts', '|', 'Average Gift'
+                       )
+    print()
+    print(title_line_form.format(*title_line_text))
+    print('- ' * 38)
+    form_line = "{:<26}{:>3}{:>13}{:>3}{:>13}{:>3}{:>13}"
+    for name in sort_donors_by_total():
+        total = get_total_given(name)
+        num_gifts = len(get_donations(name))
+        mean = round((total / num_gifts), 2)
+        print(form_line.format(str(name), '$', str(total), ' ',
+                               str(num_gifts), '$', str(mean)
+                               )
+              )
+    print()
+
+
+# PRINT ON SREEN A THANK YOU LETTER TO SOMEONE WHO JUST MADE A DONATION
+def input_donation(name):
+    """Obtain the donation amount from the user."""
+    prompt_amount = "Enter the donation amount or 0 to abort > "
+    while True:
+        try:
+            donation_amount = float(input(prompt_amount))
+        except ValueError:
+            print("Input must be a number")
+            continue
+        else:
+            if donation_amount == 0.0:
+                return False
+            else:
+                add_donation(name, donation_amount)
+                return True
+    return False
+
+
+def existing_donor_interaction():
+    """Ask for old donor name, donation amount, print a thank-you email."""
+    prompt_name = "Type full name of the old donor or 0 to abort > "
+    old_donor_name = input(prompt_name)
+    if old_donor_name == "0":
+        return
+    while old_donor_name not in donors:
+        old_donor_name = input(prompt_name)
+        if old_donor_name == "0":
+            return
+
+    if input_donation(old_donor_name):
+        print_email(old_donor_name, get_last_donation(old_donor_name))
+
+
+def new_donor_interaction():
+    """Ask for new donor name, donation amount, print a thank-you email."""
+    prompt_name = "Type full name of the new donor or 0 to abort > "
+    new_donor_name = input(prompt_name)
+    if new_donor_name == "0":
+        return
+
+    if input_donation(new_donor_name):
+        print_email(new_donor_name, get_last_donation(new_donor_name))
```

Comment:
The new donor name can old donor name appear to be the same function.  All that differs is the name.  Could you simplify this to one function?

Reply:
I think you are referring to the functions existing_donor_interaction() and new_donor_interaction(). They look similar but differ in one important behavior -- the old donor function rejects any name that the user types which is not on the old donors list, while the new donor function accepts any name typed by the user and adds such name to the donors list. The intention was to ask the user to select first if he wanted to send a letter to an old or to a new donor. If the user chooses an old donor then the program forces the user to type an existing name. If the user chooses a new donor, then any name he types suits. If I were to combine these 2 options into one option, then the user would have to be very careful if he wanted to type an existing name since any typo would lead to the creation of a new donor.

Reply:
I still think you can do this in one function.  Here is a hint, add a flag (read boolean) that states whether or not you want old_donor or new_donor functionality.

Reply:
Do you mean anything like below? (I added 2 'umbrella' functions which would be called from the dispatch menu and which would in their turn call a consolidated donor function with the right flags)
def new_donor():
    any_donor(old=False)

def old_donor():
    any_donor()

def any_donor(old=True):
    if old:
        print("Old donor code goes here")
    else:
        print("New donor code goes here")

    print("Some final code goes here")